base/compiler/types.jl

```diff
@@ -451,9 +451,12 @@ abstract type CallInfo end
 nsplit(info::CallInfo) = nsplit_impl(info)::Union{Nothing,Int}
 getsplit(info::CallInfo, idx::Int) = getsplit_impl(info, idx)::MethodLookupResult
 getresult(info::CallInfo, idx::Int) = getresult_impl(info, idx)
+add_uncovered_edges!(edges::Vector{Any}, info::CallInfo, @nospecialize(atype)) = add_uncovered_edges_impl(edges, info, atype)
+
 
 nsplit_impl(::CallInfo) = nothing
 getsplit_impl(::CallInfo, ::Int) = error("unexpected call into `getsplit`")
 getresult_impl(::CallInfo, ::Int) = nothing
+add_uncovered_edges_impl(edges::Vector{Any}, info::CallInfo, @nospecialize(atype)) = nothing
```

Comment:
With this default implementation, for any arbitrary custom `CallInfo`, as long as it correctly implements `getsplit_impl`, the method error case inlining enabled by this PR will occur. However, if that `CallInfo` does not implement `add_uncovered_edges_impl`, there could theoretically be cases where the necessary edges are not added. In the version I proposed, the `getmatchinfo(::CallInfo, ::Int)` interface is enforced to be implemented for all `CallInfo`s, so I think such interface requirement errors are easy to detect. That said, this is a very rare case, so I think it’s just fine to revisit it later when there’s more time.

Reply:
> The `getmatchinfo(::CallInfo, ::Int)` isn't quite right, because it treats uncovered edges as being associated with split indices, but they are the edge(s) not covered by the split cases (e.g. you might have 10 split cases, and then an uncovered `atype` that needs an edge to 2 method tables)

Right... So maybe `getuncoveredmt(::CallInfo) -> iterator of method tables with !fullmatch`?

Reply:
> why we provide a default implementation for any of these

Since they are optional. If `CallInfo` implements `nsplit`, then the inlining for the call is enabled and the `CallInfo` is required to implement `getsplit`. And `getresult` is optional, and if it's implemented, then the inliner may use a result from extended lattice inference (a.k.a. const-prop and semi-concrete interpretation).

Reply:
That's helpful - sounds like this one should be:
```julia
add_uncovered_edges_impl(edges::Vector{Any}, info::CallInfo, @nospecialize(atype)) = error("unexpected call into add_uncovered_edges!")
```

since we also only do this if you implement `nsplit` 

That said, `getuncoveredmt` would also be OK w/ me - the API shape seems orthogonal to the optionality, etc.

Reply:
I intend for us to eliminate all this tracking code for better correctness, as soon as someone can help finish https://github.com/JuliaLang/julia/pull/54894